### DISCUSSION TITLE
Multiple stores/groups manageament

### DISCUSSION THREAD

**Discussion by llagerlof** [gopasspw/gopass#2163] [Q&A]
Is it possible to use gopass to store groups secrets?

Best I can do to explain is to present my use case:

- Company (must have a central user administrator that can't have access to the users secrets but need to manage the git repositories to create new groups)
  - Group A (can have multiple users and all of them can manipulate and view secrets of your own group)
    - Anna (can view and manipulate secrets created by herself, John and Carl, but should not have any permission in other groups)
    - John
    - Carl
  - Group B
    - Beth
  - Group C
    - Robert
    - Jack

**Comment by dominikschulz** [comment]
This is exactly what the mounts feature of gopass was designed for.

We don't encrypt the names of secrets so you might not want to make your password store repos completley public.
But the content of the secrets should be fully protected as long as your GPG/Age setup is not compromised.

We have initially used gopass in the exact same setting like you describe above.

**Comment by AnomalRoil** [comment]
This should be done at the Git level, in my opinion, but you cannot prevent the company administrator to add their own public key to any store if they have access (but an user could notice an extra key being added during sync).

Here is the way I would set it up:
- one git repo per group: group-a / group-b / group-c
- each user has rights only on their group-a / group-b / group-c git repository
- the admin can add the public keys to the .public-keys folder of each repo and should give each user the proper git repo access rights.

Example commands to achieve this using age (with GPG it would be the same, and you could use yubikeys or hardware token for extra security):

```
[This step is in case you want to work on a dedicated folder is you also use gopass for yourself]
$ export GOPASS_HOMEDIR=/home/anomalroil/gopass-test

$  gopass setup --crypto age --create
[...]

$  gopass init --store group-a --crypto age
[... you can reuse your key here, no need to create a new one ]

$   gopass init --store group-b --crypto age
[...]

$  gopass mounts
gopass (/home/anomalroil/gopass-test/.local/share/gopass/stores/root)
├── group-a (/home/anomalroil/gopass-test/.local/share/gopass/stores/group-a)
└── group-b (/home/anomalroil/gopass-test/.local/share/gopass/stores/group-b)

$ cd /home/anomalroil/gopass-test/.local/share/gopass/stores/group-a

$ git remote add origin url-to-your-git-repo-group-a

$ cd /home/anomalroil/gopass-test/.local/share/gopass/stores/group-b
$ git remote add origin url-to-your-git-repo-group-b

[ask users to create their own local store (using gopass setup --crypto age --create) and to give you their public key]

$  gopass recipients add age1u0j5kj2a2776gphs5s7la9zzyfnaxf406kg0j43qg55ecurprfsqppplvc
[select correct mount for this recipient]

$  gopass recipients
gopass
├── age1y0lrrj0vzfpucgezu2rupz2vsraadex3m3dqdvgc58hkhy4chy0s5n5ggr
├── group-a (/home/anomalroil/gopass-test/.local/share/gopass/stores/group-a)
│   ├── age1u0j5kj2a2776gphs5s7la9zzyfnaxf406kg0j43qg55ecurprfsqppplvc (missing public key)
│   └── age1y0lrrj0vzfpucgezu2rupz2vsraadex3m3dqdvgc58hkhy4chy0s5n5ggr
└── group-b (/home/anomalroil/gopass-test/.local/share/gopass/stores/group-b)
    └── ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIHaAYBHQ9oFsThwXrPdyM+iiqYSh+EuQftWoPyE8YjM
```

@dominikschulz  BTW, any idea why it says "missing public key" here?

**Reply by dominikschulz** [reply]
That comes from how we're handling GPG. But a [fix](https://github.com/gopasspw/gopass/pull/2166) is on it's way.

**Comment by SanCoder-Q** [comment]
In our case, we'd like to centralize the stores to reduce operation load for each team and reach a certain level of consistency. Is there a way to have multiple stores in one git repo? Or is there a way to make a git folder as a "remote" of a store? e.g. `gopass git remote add --store group-a origin git@company.com/your_org/stores.git --folder group-a`

**Reply by dominikschulz** [reply]
No, I'm afraid that's not supported. This would add an amount of complexity that I'm not equipped to deal with.

Git repos are supposed to be cheap.

Do you mind elaborating how a single repo would it simpler for you?

**Reply by AnomalRoil** [reply]
Couldn't you be using git submodules for that?

**Reply by SanCoder-Q** [reply]
Thanks!

Git submodules is an option right now. It could provide a single place to manage all the secrets while each team could have their own store. It doesn't hide the fact that we need multiple git repositories for storing the data, which seems more or less a workaround to me (since it leverages the access control from git instead have it's own). 

This would be a bit harder for us since we have the limit to only use one git organization to manage all the repos unfortunately. Contrary to multi-repo solution, single repo would be easier for us to organize the secrets and do access control along with the teams' scaling. 

But thanks again for the replies. We will try the submodule solution and see how it can fit.
